netket/vqs/mc/mc_state/state.py

```diff
@@ -297,6 +302,17 @@ def sampler(self, sampler: Sampler):
         self.sampler_state = self.sampler.init_state(
             self.model, self.variables, seed=self._sampler_seed
         )
+
+        # Update `n_samples`, `n_samples_per_rank`, and `chain_length` according
+        # to the new `sampler.n_chains`.
+        # If `n_samples` is divisible by the new `sampler.n_chains`, it will be
+        # unchanged; otherwise it will be rounded up.
+        # If the new `n_samples_per_rank` is not divisible by `chunk_size`, a
+        # `ValueError` will be raised.
+        # `self.n_samples == 0` means that this `MCState` is being constructed.
+        if self.n_samples > 0:
```

Comment:
why if?

Reply:
When constructing a `MCState` in `__init__`, we first initialize `self.sampler`, then initialize `self.n_samples`. If we set `self.n_samples` inside `sampler.setter`, we will trigger the computation in `n_samples.setter` while `self.sampler` is not actually set, and that will cause problems.